### PR TITLE
Update running t-route through BMI functions with latest hydrofabric

### PR DESCRIPTION
**Pull request by shorvath-noaa**
Updates have been made to handle the most recent version of the hydrofabric. t-route can be run from either the command line or through BMI functions. Waterbodies can now be simulated, and DA for streamflow nudging and reservoir persistence can be performed if running from the command line (BMI implementation of DA will be the next task). Preprocessing of network data in HYFeaturesNetwork object has been updated to be more modular. These preprocessing steps have been generalized to run the same from either command line or BMI functions. The only difference between those two approaches now is how the raw data tables from the geopackage are loaded into the network object.

## Additions

- Modular functions for preprocessing network data (from geopackage tables)
- Updates to BMI variables (names and initialization values)
- Updates to bmi_parameters that are loaded from .yaml file
- Gages dictionary is created from 'network' table in geopackage
- Wbody_conn is created from 'flowpath_attributes' table
- bandaid() function added to account for some segment IDs that lie within waterbodies but do not have the waterbody IDs listed in 'flowpath_attribtues' table
- duplicate_wb_segments added to .yaml file. These are IDs that are shared by a segment and a waterbody. These are treated similar to synthetic_wb_segments in the NHDNetwork.
- Some hacky adjustments to creating the wbody_conn dictionary, link_lake_df, and reservoir_usgs_df. These should be updated at some point.

## Removals

-

## Changes

-

## Testing

1.

## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by shorvath-noaa** [comment]
@kumdonoaa You should be able to run the CONUS test case with test/ngen/test_conus_AnA.yaml. From command line streamflow nudging and persistence DA can be turned on. From BMI they should be turned off.

File: src/bmi_troute.py [hunk 1]
```diff
@@ -134,10 +134,10 @@ def initialize(self, bmi_cfg_file=None):
...
         bmi_cfg_file = Path(bmi_cfg_file)
         bmi_parameters = _read_config_file(bmi_cfg_file)
-        n_segment = int(bmi_parameters.get('segment_number'))
-        n_waterbody = int(bmi_parameters.get('waterbody_number'))
-        n_io = int(bmi_parameters.get('io_number'))
-        n_upstream = int(bmi_parameters.get('upstream_number'))
+        flowpath_columns = bmi_parameters.get('flowpath_columns')
+        attributes_columns = bmi_parameters.get('attributes_columns')
+        lakes_columns = bmi_parameters.get('waterbody_columns', [])
+        network_columns = bmi_parameters.get('network_columns', [])
```
Comment (kumdonoaa) [inline src/bmi_troute.py:140]: test

File: src/troute-network/troute/HYFeaturesNetwork.py [hunk 2]
```diff
@@ -497,3 +603,117 @@ def read_file(file_name):
...
+                subset_dict = {key: value for key, value in connections.items() if key in wbody_nodes}
+                link_lake[wbody_code] = list(tailwaters(subset_dict))[0]
+
+        elif reservoir_boundary(connections, waterbodies, n):
+            # one of the children of n is a member of a waterbody
+            # replace that child with waterbody code.
+            new_conn[n] = []
+
+            for child in connections[n]:
+                if child in waterbodies:
```
[the remark this answers is not in the document]
Comment (shorvath-noaa) [reply src/troute-network/troute/HYFeaturesNetwork.py:711]: Yeah, it looks through all segments. This code snippet was copied directly from [nhd_network.py](https://github.com/NOAA-OWP/t-route/blob/f3d5b1b1fcc2de8d80781dfa118a7777f70fd4d5/src/troute-network/troute/nhd_network.py#L662-L666), so this is the way t-route has been running all along.

File: src/troute-network/troute/HYFeaturesNetwork.py [hunk 3]
```diff
@@ -287,114 +360,134 @@ def read_geo_file(self,):
...
+            # clear out missing values
+            gages_df = gages_df[~gages_df['hl_uri'].isnull()]
+            gages_df = gages_df[~gages_df['hydroseq'].isnull()]
+            # make 'id' an integer
+            gages_df['id'] = gages_df['id'].str.split('-',expand=True).loc[:,1].astype(float).astype(int)
+            # split the hl_uri column into type and value
+            gages_df[['type','value']] = gages_df.hl_uri.str.split('-',expand=True,n=1)
+            # filter for 'Gages' only
+            gages_df = gages_df[gages_df['type'].isin(['Gages','NID'])]
+            # assign only the furthest downstream segment ID to any given gage
```
Comment (kumdonoaa) [inline src/troute-network/troute/HYFeaturesNetwork.py:434]: To be much clearer on this comment, we may say "among multiple flowpaths, if any, linked to a given gage by hydroseq or hl_uri attribute, assign only the furthest downstream flowpath ID to the gage"?

Comment (shorvath-noaa) [reply src/troute-network/troute/HYFeaturesNetwork.py:434]: Other way around, multiple gage IDs assigned to single flowpath ID

Comment (shorvath-noaa) [reply src/troute-network/troute/HYFeaturesNetwork.py]: Added a clarifying comment in commit [bfec9e5](https://github.com/NOAA-OWP/t-route/pull/622/commits/bfec9e5675f212cbfcb091bbe729b8e46c018a07)

File: src/troute-network/troute/HYFeaturesNetwork.py [hunk 4]
```diff
@@ -287,114 +360,134 @@ def read_geo_file(self,):
...
+            gages_df = gages_df.sort_values(['value','hydroseq']).groupby('value').last().reset_index()
+            # transform dataframe into a dictionary
+            self._gages = gages_df[['id','value']].rename(columns={'value': 'gages'}).set_index('id').to_dict()
```
[the remark this answers is not in the document]
Comment (shorvath-noaa) [reply src/troute-network/troute/HYFeaturesNetwork.py:438]: Added a clarifying comment in commit [bfec9e5](https://github.com/NOAA-OWP/t-route/pull/622/commits/bfec9e5675f212cbfcb091bbe729b8e46c018a07)

File: src/troute-network/troute/HYFeaturesNetwork.py [hunk 5]
```diff
@@ -287,114 +360,134 @@ def read_geo_file(self,):
...
+            gages_df = gages_df[~gages_df['hl_uri'].isnull()]
+            gages_df = gages_df[~gages_df['hydroseq'].isnull()]
+            # make 'id' an integer
+            gages_df['id'] = gages_df['id'].str.split('-',expand=True).loc[:,1].astype(float).astype(int)
+            # split the hl_uri column into type and value
+            gages_df[['type','value']] = gages_df.hl_uri.str.split('-',expand=True,n=1)
+            # filter for 'Gages' only
+            gages_df = gages_df[gages_df['type'].isin(['Gages','NID'])]
+            # assign only the furthest downstream segment ID to any given gage
+            gages_df = gages_df.sort_values(['value','hydroseq']).groupby('value').last().reset_index()
```
Comment (kumdonoaa) [inline src/troute-network/troute/HYFeaturesNetwork.py:436]: Does this code at L435 guarantee to find a gage ID that is furthest downstream among, if any, multiple flowpaths linked to the same gage?

File: src/troute-network/troute/HYFeaturesNetwork.py [hunk 6]
```diff
@@ -287,114 +360,134 @@ def read_geo_file(self,):
...
+            # it is the gage furthest downstream
+            lake_ids = lake_gage_df.lake_id.unique()
+            lake_outflow_gage_ids = []
+
+            for i in lake_ids:
+                temp_df = lake_gage_df[lake_gage_df['lake_id']==i]
+                lake_outflow_gage_ids.append(
+                    gages_df[gages_df['value']
+                             .isin(temp_df.gages)]
+                             .sort_values('hydroseq')
```
Comment (kumdonoaa) [inline src/troute-network/troute/HYFeaturesNetwork.py:466]: Does this sort_values() by 'hydroseq' guarantee to find a gage that is furthest downstream among, if any, multiple gages linked to the same watebody?

File: src/troute-network/troute/HYFeaturesNetwork.py (hunk 3 above)
Comment (shorvath-noaa) [reply src/troute-network/troute/HYFeaturesNetwork.py]: I'm mistaken, this code does indeed find the furthest downstream segment when there are multiple segments assigned to a single gage ID

File: src/bmi_troute.py (hunk 1 above)
Comment (shorvath-noaa) [reply src/bmi_troute.py:140]: ?

**Review by kumdonoaa — APPROVED** [review]

**Pull request merged**